### PR TITLE
Add unit tests for ITypeOfExpression and ISizeOfExpression and make A…

### LINKED ISSUE #22003: API review for ITypeOfExpression
https://github.com/dotnet/roslyn/blob/features/ioperation/src/Compilers/Core/Portable/Operations/ITypeOfExpression.cs

Add units tests when making public.

### PR DESCRIPTION
**Pull request by mavasani**
…PIs public again

Fixes #22003 and #21296

### CODE REVIEW

File: Compilers.sln [hunk 1]
```diff
@@ -159,7 +158,6 @@ Global
 		src\Compilers\Core\CommandLine\CommandLine.projitems*{4b45ca0c-03a0-400f-b454-3d4bcb16af38}*SharedItemsImports = 4
 		src\Dependencies\PooledObjects\Microsoft.CodeAnalysis.PooledObjects.projitems*{5f8d2414-064a-4b3a-9b42-8e2a04246be5}*SharedItemsImports = 4
 		src\Compilers\Core\CommandLine\CommandLine.projitems*{7ad4fe65-9a30-41a6-8004-aa8f89bcb7f3}*SharedItemsImports = 4
-		src\Compilers\Core\CommandLine\CommandLine.projitems*{8ce3a581-2969-4864-a803-013e9d977c3a}*SharedItemsImports = 4
```
Comment (jcouv) [inline Compilers.sln]: What are those changes?

Comment (mavasani) [reply Compilers.sln]: Ah, did not intend to check this in, let me revert.

File: src/Compilers/Core/Portable/PublicAPI.Unshipped.txt [hunk 2]
```diff
@@ -328,8 +330,8 @@ Microsoft.CodeAnalysis.Semantics.IIsPatternExpression
 Microsoft.CodeAnalysis.Semantics.IIsPatternExpression.Expression.get -> Microsoft.CodeAnalysis.IOperation
 Microsoft.CodeAnalysis.Semantics.IIsPatternExpression.Pattern.get -> Microsoft.CodeAnalysis.Semantics.IPattern
 Microsoft.CodeAnalysis.Semantics.IIsTypeExpression
-Microsoft.CodeAnalysis.Semantics.IIsTypeExpression.IsType.get -> Microsoft.CodeAnalysis.ITypeSymbol
 Microsoft.CodeAnalysis.Semantics.IIsTypeExpression.IsNotTypeExpression.get -> bool
```
Comment (jcouv) [inline src/Compilers/Core/Portable/PublicAPI.Unshipped.txt]: Anyone know why those unrelated changes are happening? I assume this file is maintained in a sorted order, but it looks like the previous checked-in version was not.

Comment (mavasani) [reply src/Compilers/Core/Portable/PublicAPI.Unshipped.txt]: Likely that some previous change was hand crafted and not via code fix, and it broke the ordering.

Comment (mavasani) [reply src/Compilers/Core/Portable/PublicAPI.Unshipped.txt]: I have also seen merge resolution introduce such ordering issues.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ITypeOfExpression.cs [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+        public void TestTypeOfExpression()
+        {
+            string source = @"
+using System;
+
+class C
+{
+    void M(Type t)
+    {
+        t = /*<bind>*/typeof(int)/*</bind>*/;
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ITypeOfExpression.cs]: Will VB's `GetType` be represented as a typeof IOperation? If not, do we know how it will be represented?

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ISizeOfExpression.cs [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+                Diagnostic(ErrorCode.ERR_TypeExpected, ")").WithLocation(8, 30),
+                // CS0233: '?' does not have a predefined size, therefore sizeof can only be used in an unsafe context (consider using System.Runtime.InteropServices.Marshal.SizeOf)
+                //         i = /*<bind>*/sizeof()/*</bind>*/;
+                Diagnostic(ErrorCode.ERR_SizeofUnsafe, "sizeof()").WithArguments("?").WithLocation(8, 23)
+            };
+
+            VerifyOperationTreeAndDiagnosticsForTest<SizeOfExpressionSyntax>(source, expectedOperationTree, expectedDiagnostics);
+        }
+    }
+}
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ISizeOfExpression.cs]: There is no VB equivalent of `sizeof` to test, correct?

File: Compilers.sln (hunk 1 above)
Comment (khyperia) [reply Compilers.sln]: I may have screwed this up in https://github.com/dotnet/roslyn/pull/22142/files ... actually yes, the `SharedItemsImports` deletion is correct, as the two GUIDs deleted in this PR are CscCore and VbcCore.

I'm not sure about the `VisualStudioVersion` change or the `GlobalSection(ExtensibilityGlobals)` addition.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ITypeOfExpression.cs (hunk 3 above)
Comment (mavasani) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ITypeOfExpression.cs]: I thought we had a separate API for `GetType`, but apparently we do not have it. I agree that it makes sense for `GetType` to be represented as an `ITypeOfExpression`. Tagging @dotnet/analyzer-ioperation for confirmation.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ISizeOfExpression.cs (hunk 4 above)
Comment (mavasani) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ISizeOfExpression.cs]: Yes, I could not find any.

**Review by jcouv — APPROVED** [review]
LGTM with follow-up on VB GetType.

**Review by AlekseyTs — APPROVED** [review]
LGTM

**Comment by mavasani** [comment]
Flaky test failure: https://github.com/dotnet/roslyn/issues/22230

**Comment by mavasani** [comment]
@dotnet-bot retest windows_debug_unit32_prtest please
